### PR TITLE
Get OpenAI working

### PR DESCRIPTION
**Pull request by hunterjackson**

### CODE REVIEW

File: src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    int totalTokens = functionTokens;
+    totalTokens += 3; // every reply is primed with <|start|>assistant<|message|>
+
+    JsonNode systemMessage = messages.get(0);
+    boolean hasSystemMessage = systemMessage.get("role").textValue().equals("system");
+    if (hasSystemMessage) {
+      // if the system message is present it's required
+      totalTokens += tokenCount(messages.get(0));
+    }
+    for (int i = messages.size() - 1; i >= 0; i--) {
```
Comment (colinmccloskey) [inline src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java:105]: Not necessarily cleaner but we can remove the above if statement on 101 and the continue on 109 if its for (int i = hasSystemMessage ? 0 : int i = messages.size() - 1; i >= hasSystemMessage ? -messages.size() + 2 : 0; i--)

Comment (colinmccloskey) [reply src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java:105]: Would also need to change the input on 115/118 though so maybe not worth it

File: src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+      messages
+          .addObject()
+          .put("role", message.role().toString().toLowerCase())
+          .put("content", message.message());
+    }
+
+    Optional<ArrayNode> prunedMessages = pruneMessages(messages, null);
+    if (prunedMessages.isEmpty()) {
+      return messageStack.newMessageFromBot(
+          Instant.now(), "I'm sorry but that request was too long for me.");
```
Comment (colinmccloskey) [inline src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java:169]: If this is getting send back to the user, maybe add something to ask for them to repeat it in fewer words?

**Review by colinmccloskey — APPROVED** [review]
shiiiip

File: src/main/java/com/meta/chatbridge/message/FBMessageHandler.java [hunk 3]
```diff
@@ -73,11 +74,11 @@ public FBMessageHandler(String verifyToken, String pageAccessToken, String appSe
     this.accessToken = pageAccessToken;
   }
 
-  @TestOnly
-  @This
-  FBMessageHandler baseURLFactory(Function<Identifier, URI> baseURLFactory) {
-    this.baseURLFactory = Objects.requireNonNull(baseURLFactory);
-    return this;
+  private static Stream<String> textChunker(String text, String regexSeparator) {
+    if (text.length() > 2000) {
```
[the remark this answers is not in the document]
Comment (hunterjackson) [reply src/main/java/com/meta/chatbridge/message/FBMessageHandler.java:78]: > why 2000?

2k is the max number of characters allowed in a messenger message

> And would be want to be removing line breaks from large chunks of text? Might be more readable that way (maybe be misunderstanding the code)

Yes it does remove the breaks, however each message is sent as an individual message so we still have breaks

File: src/main/java/com/meta/chatbridge/message/MessageStack.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    if (newMessage.timestamp().isBefore(old.tail().timestamp())) {
+      this.messages =
+          Stream.concat(messages.stream(), Stream.of(newMessage))
+              .sorted(Comparator.comparing(Message::timestamp))
+              .collect(Collectors.toUnmodifiableList());
+    } else {
+      this.messages = ImmutableList.<T>builder().addAll(messages).add(newMessage).build();
+    }
+
+    Preconditions.checkArgument(
```
[the remark this answers is not in the document]
Comment (hunterjackson) [reply src/main/java/com/meta/chatbridge/message/MessageStack.java:47]: conversation id is the two concatenated with a separator yes, but they are always applied in the same order, so can swap them and get the same conversation id.

> And silly Java question, will it also show the IDs in question if it's not a templated error message?

When you concatenate with the `+` operator it calls the `toString` function

File: src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java (hunk 2 above)
Comment (hunterjackson) [reply src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java:169]: Yea I thought about that and I want to get some feedback from users here, maybe make it customizable, I'm not sure.

File: src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java (hunk 1 above)
Comment (hunterjackson) [reply src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java:105]: yea it's definitely not an elegant way to code this, but I had trouble find an elegant way to do it

**Pull request merged**